### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Critical Bug**: Inconsistent HTML tag validation between title and description. The title regex `/<[^>]*>/g` correctly rejects all HTML tags, but the description regex `/<.*>/g` is overly permissive and would allow malformed tags like `<img` or tags with newlines. This creates a security inconsistency.

**Recommendation**: Use the same regex pattern for both fields:
```typescript
else if (/<[^>]*>/g.test(description)) {
```

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
+  if (!id || !/^\d+$/.test(id)) {
+    return res.status(400).json({
+      error: 'Invalid ID format',
+      details: [{ field: 'id', message: 'ID must be a positive integer', value: id }]
+    });
+  }
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:109]: **Security Issue**: The current ID validation allows `"0"` which may not be valid depending on your ID generation strategy. Most auto-incrementing IDs start from 1.

**Recommendation**: Consider using a more restrictive pattern if ID 0 is not valid:
```typescript
if (!id || !/^[1-9]\d*$/.test(id)) {
```
This ensures only positive integers starting from 1 are accepted.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Reliability Issue**: Missing error handling around `todoStore.create()`. If this method throws an exception (e.g., due to storage issues, validation errors, or memory constraints), the server will crash.

**Recommendation**: Add try-catch block:
```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  res.status(500).json({ 
    error: 'Internal server error',
    details: [{ field: 'server', message: 'Failed to create todo' }]
  });
}
```

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **Test Coverage Gap**: This comment indicates missing test coverage for description HTML validation. Given the regex inconsistency bug I found in the validation logic, this test would have caught the issue.

**Recommendation**: Add this test case:
```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    it('should reject non-numeric ID', async () => {
+      const response = await request(app).get('/abc');
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'id',
+        message: 'ID must be a positive integer',
+        value: 'abc'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:119]: **Test Coverage Gap**: This missing test for ID "0" edge case is important given the validation bug I identified. The current implementation would accept "0" as a valid ID.

**Recommendation**: Add this test case:
```typescript
it('should reject ID "0" if not valid', async () => {
  const response = await request(app).get('/0');
  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'id',
    message: 'ID must be a positive integer',
    value: '0'
  });
});
```

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Code Review Summary

This PR adds comprehensive input validation middleware to the Todo API, which is a significant improvement for security and data integrity. The implementation demonstrates good understanding of validation patterns and includes extensive test coverage. However, I've identified several critical issues that need to be addressed before approval.

### ✅ **Strengths**

1. **Comprehensive Validation**: Excellent coverage of input validation scenarios including type checking, length limits, and XSS prevention
2. **Structured Error Responses**: The new error format with field-specific details is much more user-friendly than simple error strings
3. **Good Test Coverage**: 15+ test cases covering various validation scenarios
4. **Security Focus**: HTML tag prevention to mitigate XSS attacks is well-implemented
5. **Clean Architecture**: Separation of validation logic into middleware is good practice
6. **Documentation**: README updates clearly explain the validation rules and new error format

### ⚠️ **Critical Issues**

1. **Security Bug**: Inconsistent HTML regex patterns between title and description validation could allow malformed HTML tags through description field
2. **ID Validation Gap**: Accepts ID "0" which may not be valid for your ID system
3. **Missing Error Handling**: No try-catch around `todoStore.create()` could crash the server
4. **Test Coverage Gaps**: Missing tests for description HTML validation and ID edge cases (the author's own comments confirm this)

### 🔧 **Minor Issues**

- Inconsistent regex patterns make the code harder to maintain
- The `ValidationException` class is defined but never used
- Some validation logic is duplicated between create and update validators

### 📋 **Breaking Changes Acknowledged**

The error response format change is properly documented as a breaking change, which is good practice.

### 🚀 **Recommendations**

1. Fix the HTML regex inconsistency (critical)
2. Review ID validation logic for "0" case
3. Add error handling around storage operations
4. Add the missing test cases mentioned in the comments
5. Consider extracting common validation patterns into reusable functions

This is solid work that significantly improves the API's robustness. Once the critical issues are addressed, this will be ready for merge.

**Overall Assessment**: Request Changes - Great foundation but needs critical bug fixes before merge.
